flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateDownloader.java

```diff
@@ -90,27 +80,62 @@ private void downloadDataForAllStateHandles(
             } else {
                 throw new FlinkRuntimeException("Failed to download data for state handles.", e);
             }
+        } finally {
+            // Unregister and close the internal closer. In a failure case, this should interrupt
+            // ongoing downloads.
+            closeableRegistry.unregisterCloseable(internalCloser);
+            IOUtils.closeQuietly(internalCloser);
+            if (failureCleanupRequired) {
+                // Cleanup on exception: cancel all tasks and delete the created directories
+                futures.forEach(future -> future.cancel(true));
+                downloadRequests.stream()
+                        .map(StateHandleDownloadSpec::getDownloadDestination)
+                        .map(Path::toFile)
+                        .forEach(FileUtils::deleteDirectoryQuietly);
+            }
```

Comment:
Why can't we just move error handling logic to `catch` and remove `failureCleanupRequired`?

Reply:
Two reasons:
- the code was written to catch only execution exception, but I want to make sure the cleanup happens also for any other exception.
- For cleanup, I first want to close the registry to interrupt the IO in all parallel tasks, but I also want to close the registry in the finally block, so it would have slightly more code duplication.

Reply:
1. Why not to change the exception type in the `catch` block, so that it works for any type of exception? I don't think the intention there was to catch **only** `ExecutionException`
2. Doesn't `future.cancel(true)` interrupt the IO?

Reply:
1. But the special handling in the catch block is only meant for ExecutionException so we need to check separately etc.
2. That is not guaranteed and closing the streams is the safe way.

Reply:
About 1, we'd have to ask the original author, but the code calls `ExceptionUtils.stripExecutionException(e);` , so it's expecting to work on execution exceptions. Not that it couldn't be changed...

And for 2, yes that can be done, but is it actually better or more readable? I think we are debating about personal taste here, except if you have a point while one should be better than the other way?

Reply:
Yes, I think that would be more readable without conditional logic, but I'd be fine with either option.

>  if you have a point while one should be better than the other way

> That way, all code that should be executed anyways lives in `finally`, and error handling - in `catch`.